### PR TITLE
[docker] Remove cached git repositories from docker image

### PR DESCRIPTION
**Pull request by WonyoungChoi**
Use `gclient-prepare-sync` script newly added instead of the `.git` cached directory for reducing `gclient sync` time.

In previous PR (#176 ), the build docker image had some cached `.git` directories for `gclient sync` time shortening, which was not a great help because of increasing download time of the docker image. 😢 

Therefore, I offer a new method.
- Make a simple docker image based on `tizen-tools` image.
- Add  `gclient-prepare-sync` script to shorten `gclient-sync` time. The `gclient sync` downloads overly many objects to perform builds, even if the `--no-history` option is used. This script downloads only single commit required for the build to shorten the overall `gclient sync` time.
- And some scripts are added to make the build worklfow simply. 

Test result: https://github.com/WonyoungChoi/engine/actions/runs/1197699330
Test worklfow: https://github.com/WonyoungChoi/engine/blob/citest/.github/workflows/build.yml

### the sample build workflow:
```yaml
    env:
      CACHE_PATH: src/out/linux_${{ matrix.mode }}_${{ matrix.arch }}
      CACHE_PREFIX: out-build-linux_${{ matrix.mode }}_${{ matrix.arch }}

    steps:
    - uses: actions/checkout@v2
      with:
        path: src/flutter

    - uses: actions/cache@v2
      with:
        path: ${{ env.CACHE_PATH }}
        key: ${{ env.CACHE_PREFIX }}-${{ github.sha }}
        restore-keys: |
          ${{ env.CACHE_PREFIX }}-

    - name: gclient sync
      run: |
        gclient-prepare-sync.sh --reduce-deps --shallow-sync
        gclient sync -v --no-history --shallow

    - name: prepare incremental build
      run: |
        cache-checksum.sh restore $CACHE_PATH

    - name: build
      run: |
        build-engine.sh ${{ matrix.mode }} ${{ matrix.arch }} ${{ matrix.triple }}

    - name: save checksum
      run: |
        cache-checksum.sh save $CACHE_PATH

```

### CODE REVIEW

File: ci/docker/tizen/tools/cache-restore-mtime [hunk 1]
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env bash
+# Copyright 2021 Samsung Electronics Co., Ltd. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+# After `gclient sync`, the mtime of all synchronized files will be changed
+# to newer than the cached out files. Therefore, this script changes the mtime
+# of all files to OLD time and changes only the mtime of the changed files
+# to NEW time.
```
Comment (swift-kim) [inline ci/docker/tizen/tools/cache-restore-mtime:9]: Why does this matter?

Actually the overall workflow is not very intuitive for me. For what do you restore mtimes and save checksums of files? Are both `gclient-reduce-deps` and `gclient-fast-sync` required and should they be separate files? (As far as I understand the both exist for the same reason: reducing the gclient sync time.) Why do you run `gclient sync -v --no-history --shallow` after `gclient-fast-sync`? (Isn't `gclient-fast-sync` supposed to replace `gclient sync`, assuming from its name?) Why is `gclient-fast-sync.py` alone a python script while other tools are written in shell script?

**Review by swift-kim — COMMENTED** [review]
> In previous PR (#176 ), the build docker image had some cached `.git` directories for `gclient sync` time shortening, which was not a great help because of increasing download time of the docker image. 😢 

Why does caching the directories increase the image downloading time? As far as I understand they are fake git directories and supposed to be small in size.

**Comment by WonyoungChoi** [comment]
> > In previous PR (#176 ), the build docker image had some cached `.git` directories for `gclient sync` time shortening, which was not a great help because of increasing download time of the docker image. 😢
> 
> Why does caching the directories increase the image downloading time? As far as I understand they are fake git directories and supposed to be small in size.

In previous PR (#176 ), the cached .git directories are about dart, dart's third parties and skia. that was about 2GB size.

File: ci/docker/tizen/tools/build-engine [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+sed -i 's/"-Wno-non-c-typedef-for-linkage",//g' src/build/config/compiler/BUILD.gn
+sed -i 's/"-Wno-psabi",//g' src/build/config/compiler/BUILD.gn
+
+# Run gn
+src/flutter/tools/gn \
+    --target-os linux \
+    --linux-cpu $BUILD_ARCH \
+    --no-goma \
+    --target-toolchain "$TIZEN_TOOLS_PATH"/toolchains \
+    --target-sysroot "$TIZEN_TOOLS_PATH"/sysroot/$BUILD_ARCH \
```
[the remark this answers is not in the document]
Comment (WonyoungChoi) [reply ci/docker/tizen/tools/build-engine:35]: TIZEN_TOOLS_PATH can be from environment variables. I just considered "white spaces" in the variable.

File: ci/docker/tizen/tools/cache-restore-mtime (hunk 1 above)
Comment (WonyoungChoi) [reply ci/docker/tizen/tools/cache-restore-mtime:9]: 1. The reason for `restoring mtimes` is for the incremental build. After `gclient sync`, the mtime of all synchronized files will be changed to newer than the cached out files. So ninja build will try to clean build always. to solve this problem, this script changes the mtime  of all files to OLD time except new changed files.
2. Yes, those scripts can be one. But I was difficult to set the name of that script....
3. Please see the test workflow file: https://github.com/WonyoungChoi/engine/blob/citest/.github/workflows/build.yml. 
```yaml
    - name: gclient sync
      run: |
        gclient-reduce-deps
        gclient-fast-sync
        gclient sync -v --no-history --shallow
```
4. `gclient-fast-sync.py` script uses some python classes of the `depot-tools` to parse DEPS file.

File: ci/docker/tizen/tools/build-engine [hunk 3]
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (WonyoungChoi) [reply ci/docker/tizen/tools/build-engine:1]: Well.. It is just gclient style.. 😄 Do you prefer .sh suffix?

File: ci/docker/tizen/tools/cache-restore-mtime (hunk 1 above)
Comment (WonyoungChoi) [reply ci/docker/tizen/tools/cache-restore-mtime:9]: how about `gclient-prepare-sync` for the merged one. looks like.. 
```yaml
    - name: gclient sync
      run: |
        gclient-prepare-sync --reduce-deps --shallow-sync
        gclient sync -v --no-history --shallow
```

**Review by swift-kim — APPROVED** [review]

**Pull request merged**